saleor/order/migrations/0181_order_subtotal_as_a_field.py

```diff
@@ -0,0 +1,67 @@
+from django.db import migrations
+from django.db.models import F, Func, OuterRef, Subquery
+
+# One batch took 0.1s and 10.54MB at its peak
+BATCH_SIZE = 1000
+
+
+def update_order_subtotals(apps, schema_editor):
+    Order = apps.get_model("order", "Order")
+    OrderLine = apps.get_model("order", "OrderLine")
+
+    first_order = Order.objects.last()
+    if first_order is None:
+        return  # Exit if there are no orders
+    start_number = first_order.number
+
+    while True:
+        # right here we need to find only Orders which was not populated
+        # with subtotal in previous release, so the amount of rows to populate
+        # shouldn't be big
+        queryset = Order.objects.order_by("number").filter(
+            number__gte=start_number, subtotal_gross_amount__exact=0
+        )[:BATCH_SIZE]
+        current_batch_order_numbers = list(queryset.values_list("number", flat=True))
+
+        # Break the loop if there are no orders
+        if not current_batch_order_numbers:
+            break
+
+        order_line = OrderLine.objects.filter(order=OuterRef("pk"))
+        subtotal_net_sum = order_line.annotate(
+            net_sum=Func(F("total_price_net_amount"), function="Sum")
+        ).values("net_sum")
+
+        subtotal_gross_sum = order_line.annotate(
+            gross_sum=Func(F("total_price_gross_amount"), function="Sum")
+        ).values("gross_sum")
+
+        orders_with_totals = Order.objects.filter(
+            number__in=current_batch_order_numbers
+        ).annotate(
+            subtotal_net_sum=Subquery(subtotal_net_sum),
+            subtotal_gross_sum=Subquery(subtotal_gross_sum),
+        )
+
+        orders_to_update = []
+        for order in orders_with_totals:
+            order.subtotal_net_amount = order.subtotal_net_sum
+            order.subtotal_gross_amount = order.subtotal_gross_sum
+            orders_to_update.append(order)
+
+        Order.objects.bulk_update(
+            orders_to_update, ["subtotal_net_amount", "subtotal_gross_amount"]
+        )
+        start_number = current_batch_order_numbers[-1]
```

Comment:
#14932, in celery task you have:
```python
last_number = current_batch_order_numbers[-1]
update_order_subtotals.delay(last_number + 1)
```
I don't see `+1` here. Which one is correct?

Reply:
Both are correct, first one in `3.18` version is Celery based and the `main` one is non-Celery.

Non-Celery: The continuous loop retains context, and the `gte` query implicitly moves to the next unprocessed order.

Celery: Each task is independent, requiring an explicit definition of the starting point for each new task `(last_number + 1)`, ensuring no overlap and that each order is processed only once.